### PR TITLE
Bug 1838716: improve reason text on degraded condition

### PR DESCRIPTION
**Pull request by gabemontero**
explicitly cite file system errors, include apimachinery reason text on api server errors

### CODE REVIEW

**Comment by gabemontero** [comment]
Another use of upstream templates slipped through, and is not catching us on image-eco:

```
STEP: calling oc new-app https://github.com/sclorg/django-ex.git
May 22 14:16:18.694: INFO: Running 'oc --namespace=e2e-test-s2i-python-pnnbw --kubeconfig=/tmp/configfile693775026 new-app https://github.com/sclorg/django-ex.git --strategy=source -e WEB_CONCURRENCY=1'
--> Found image 7ac507f (10 days old) in image stream "openshift/python" under tag "3.6" for "python"

    Python 3.6 
    ---------- 
    Python 3.6 available as container is a base platform for building and running various Python 3.6 applications and frameworks. Python is an easy to learn, powerful programming language. It has efficient high-level data structures and a simple but effective approach to object-oriented programming. Python's elegant syntax and dynamic typing, together with its interpreted nature, make it an ideal language for scripting and rapid application development in many areas on most platforms.

    Tags: builder, python, python36, python-36, rh-python36

    * The source repository appears to match: python
    * A source build using source code from https://github.com/sclorg/django-ex.git will be created
      * The resulting image will be pushed to image stream tag "django-ex:latest"
      * Use 'oc start-build' to trigger a new build

--> Creating resources ...
    imagestream.image.openshift.io "django-ex" created
    buildconfig.build.openshift.io "django-ex" created
    deployment.apps "django-ex" created
    service "django-ex" created
--> Success
    Build scheduled, use 'oc logs -f bc/django-ex' to track its progress.
    Application is not exposed. You can expose services to the outside world by executing one or more of the commands below:
     'oc expose svc/django-ex' 
    Run 'oc status' to view your app.
STEP: waiting for build to finish
May 22 14:18:11.761: INFO: waiting for deploymentconfig e2e-test-s2i-python-pnnbw/django-ex to be available with version 1

May 22 14:18:12.797: INFO: got error "deploymentconfigs.apps.openshift.io \"django-ex\" not found" when waiting for deploymentconfig e2e-test-s2i-python-pnnbw/django-ex to be available with version 1
```

**Comment by gabemontero** [comment]
unrelated flake

/test e2e-aws

File: pkg/stub/config.go [hunk 1]
```diff
@@ -231,6 +232,19 @@ func (h *Handler) processError(opcfg *v1.Config, ctype v1.ConfigConditionType, c
...
 		status.Message = log
+
+		// most reasons for degraded status are handled in ClusterOperatorStatusDegradedCondition based on
+		// state in the opcfg correlated with other factors; but when the error should directly drive the
+		// reason text, we set the reason here; at this time, that means either API Server errors or
+		// errors accessing the local file system
+		reason := kerrors.ReasonForError(err)
+		switch {
+		case len(reason) > 0 && reason != kapis.StatusReasonUnknown:
+			status.Reason = fmt.Sprintf("APIServer%sError", reason)
```
Comment (adambkaplan) [inline pkg/stub/config.go:243]: Would it make sense to simply bubble the reason outward? In these cases would the message in `log` be sufficient?

Comment (gabemontero) [reply pkg/stub/config.go:243]: > Would it make sense to simply bubble the reason outward? In these cases would the message in `log` be sufficient?

IMO no, a log in the pod log is not sufficient.  This concise nugget in the reason facilitates triage of degraded clusters seen in OTA analysis.  And getting must gather data from customers has proven lengthy / problematic.

This provides us a low cost / low impact means of connecting sample operator degradation to issues in underlying components.

**Comment by gabemontero** [comment]
PR title / commit message updated

**Comment by gabemontero** [comment]
@mfojtik - how keen are you on getting this change backported from 4.5 to 4.4/4.3 ?

**Pull request merged**